### PR TITLE
feat: allow run test by teacher in dry-run mode

### PR DESCRIPTION
**Pull request by emgolubev**
Related to: https://oat-sa.atlassian.net/browse/TR-2715

Depends on:
- [x] https://github.com/oat-sa/extension-tao-delivery/pull/525

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -64,7 +64,7 @@
         "oat-sa/oatbox-extension-installer": "~1.1||dev-master",
 		"oat-sa/generis" : ">=14.0.0",
 		"oat-sa/tao-core" : ">=48.56.0",
-		"oat-sa/extension-tao-delivery" : ">=15.6.0",
+		"oat-sa/extension-tao-delivery" : ">=15.8.0",
```
Comment (boajer) [inline composer.json:67]: This might be changed, please ensure before merging

**Review by Businesspunk — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Commits messages are meaningful
- [x] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: model/execution/DryRunChecker.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+use common_session_SessionManager;
+use oat\taoDelivery\model\execution\DryRunCheckerInterface;
+use oat\taoLti\models\classes\LtiRoles;
+
+class DryRunChecker implements DryRunCheckerInterface
+{
+    public function isDryRun(): bool
+    {
+        $session = common_session_SessionManager::getSession();
```
Comment (boajer) [inline model/execution/DryRunChecker.php:34]: May be better to inject the session as the external dependency there?

Comment (emgolubev) [reply model/execution/DryRunChecker.php:34]: Discussed with @boajer

**Review by boajer — APPROVED** [review]
- [ ]  New code is covered by tests (if applicable)
- [ ]  Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ]  New code is respecting code style rules
- [ ]  New code is respecting best practices
- [ ]  New code is not subject to concurrency issues (if applicable)
- [ ]  Feature is working correctly on my local machine (if applicable)
- [ ]  Acceptance criteria are respected
- [ ]  Pull request title and description are meaningful
- [ ]  Pull request's target is not master
- [ ]  Commits are following conventional commits
- [ ]  Commits messages are meaningful
- [ ]  Commits are atomic
- [ ]  Changelog is updated according to changes (if applicable)
- [ ]  Documentation is updated according to changes (if applicable)

File: controller/DeliveryTool.php [hunk 3]
```diff
@@ -105,14 +104,24 @@ public function run()
...
+                throw new LtiException(__('Test Session not found'));
+            }
+
+            $user = $session->getUser();
+
             $isLearner = !is_null($user)
                 && count(array_intersect([LtiRoles::CONTEXT_LEARNER, LtiRoles::CONTEXT_LTI1P3_LEARNER], $user->getRoles())) > 0;
 
-            if ($isLearner) {
+            $isDryRun = !$isLearner && in_array(LtiRoles::CONTEXT_LTI1P3_INSTRUCTOR, $user->getRoles());
```
Comment (quintanilhar) [inline controller/DeliveryTool.php:118]: Side note: this is the type of logic we should avoid in a controller.

**Review by quintanilhar — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [X] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [X] Pull request's target is not `master`
- [X] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Commits messages are meaningful
- [ ] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: migrations/Version202111191637067335_ltiDeliveryProvider.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public function up(Schema $schema): void
+    {
+        $deliveryServerService = $this->getServiceManager()->get(DeliveryServerService::SERVICE_ID);
+
+        $deliveryServerService->setOption(
+            DeliveryServerService::OPTION_RESULT_SERVER_SERVICE_FACTORY,
+            new Lti1p3ResultServerServiceFactory()
+        );
+
+        $this->registerService(DeliveryServerService::SERVICE_ID, $deliveryServerService);
```
Comment (wazelin) [inline migrations/Version202111191637067335_ltiDeliveryProvider.php:29]: Something like that should do the trick.
```suggestion
        $this->propagate(new oat\ltiDeliveryProvider\scripts\install\RegisterLti1p3ResultServerServiceFactory())([]);
```

**Review by wazelin — CHANGES_REQUESTED** [review]

File: migrations/Version202111191637067335_ltiDeliveryProvider.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public function down(Schema $schema): void
+    {
+        throw new IrreversibleMigration(
+            'Configuration file has been modified'
+        );
```
Comment (wazelin) [inline migrations/Version202111191637067335_ltiDeliveryProvider.php:36]: It is quite reversible IMO.

```suggestion
        $deliveryServerService = $this->getServiceManager()->get(DeliveryServerService::SERVICE_ID);

        $deliveryServerServiceOptions = $deliveryServerService->getOptions();
        unset($deliveryServerServiceOptions[DeliveryServerService::OPTION_RESULT_SERVER_SERVICE_FACTORY]);

        $this->registerService(DeliveryServerService::SERVICE_ID, $deliveryServerService);
```

**Review by wazelin — CHANGES_REQUESTED** [review]

**Review by wazelin — APPROVED** [review]
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [ ] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`

**Pull request merged**